### PR TITLE
'save as draft' option for submission form #5651

### LINKED ISSUE #5651: 'save as draft' option for submission form
individual submit button for every question. Show a `Saving question ....` text above the button and change it to `Question saved at [time]` after the ajax call returns

### PR DESCRIPTION
**Pull request by tarun14110**
Fixes  #5651
screenshot from 2016-06-19 23 46 20

### CODE REVIEW

**Comment by tarun14110** [comment]
@damithc have a look over UI. The functionality is like- individual question can be submitted, till ajax response comes back label will show `Saving question` and after successful submission label will show `last submitted at ....`. 
These buttons should be named in a way such that these individual submission seems optional (I mean user may think that he have to submit every question individually before submitting the form) ?

**Comment by damithc** [comment]
Sorry Tarun, I'm having doubts about this feature again. I fear it will eventually interfere with two other future features: compulsory questions, submission receipts.

Let's consider another approach: saving drafts is an almost invisible automatic backup mechanism that user does not have any control over
* Drafts will be automatically saved periodically, if there are changes to answers
* Drafts will not be validated and visible to feedback recipients or instructors
* After the form is submitted, there will be no more draft saving

While this is not a perfect solution, I think it is a reasonable compromise that will reduce users losing lot of work due to connectivity/computer problems. What do you think? @kanghj

**Comment by kanghj** [comment]
Yes, not allowing future drafts after a submission should resolve most of the difficulty in implementing this. We should go for this. 

But the current implementation by @tarun14110 may not clash with those 2 future features (compulsory questions, submission receipts) that much, given that we already have a feature of submitting responses to only 1 question. We could reuse the submission logic for submitting to a single question for this "Submit this answer and continue". The implementation of the 2 future features (#4727, #4676) will need to consider and handle this case of a student submitting for only 1 question anyway.

**Comment by damithc** [comment]
The existing 'submit one answer' is a feature we added to fulfill requirements of a grant, not really a feature we needed. I don't think anybody is using it. Furthermore, I think that feature was implemented before we made it optional for students to use Google accounts. So the feature will not work if the student is not using the Google account. The grant requirement is no longer applicable. I think that feature should be removed :-) Shall we remove it? It's current state is half baked/broken anyway.

**Comment by kanghj** [comment]
Yes, I also think we should remove it if nobody is using it.

**Comment by tarun14110** [comment]
@damithc @kanghj should I work on autosave discussed above ? Or we are completely removing this feature ?

**Comment by tarun14110** [comment]
@damithc @kanghj I have an idea. I am not sure this will work or not. please have a look.
Autosave will happen periodically of the whole form. Before every autosave there will a kind of confirmation from the user like `would you like to submit the filled answers and continue with other answers. This will reduce the chances of losing the data due to machine or network failure`.
It will be irritating if popup comes while the user is in middle of the answer. To avoid this we can check another parameter before displaying popup i.e `if user changed the field after the periodic time ends`. example- if period is 2 min, after 2 min we will popup whenever user changes the field (starts filling other answer).

I think this will help `not to save, what user dont want to save`. And we can also reduce loss due to network or other issues ?

**Comment by damithc** [comment]
Good suggestion but I think we don't have to ask the user for confirmation. The purpose of the behind the scene 'save draft' feature is to be able to restore the submission form to it's previous state (or come as close to it as possible) in case the computer crashes or the user has to abort and resume later. Therefore, it's fine as long as the values match what is in the form. 

However, it seems this feature needs a lot of work with not so much payback. May be we just inform the student they they can save any time can continue to edit later?

**Comment by tarun14110** [comment]
just a label to inform that, you can save any time and continue later ? No need for additional functionality like Autosave ?

**Comment by damithc** [comment]
Yes. Rather than a label, we can show a dialog if the form is empty. If an 'empty form detection' is hard to do, we can show it as a info status message at the top of the form.

File: src/main/java/teammates/ui/controller/FeedbackSubmissionEditSaveDraftAction.java [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+            removeRespondant();
+        }
+        
+        String timeStamp = new SimpleDateFormat("HH.mm.ss").format(new Date());
+        
+        FeedbackSubmissionSaveDraftNotificationAjaxPageData lastSubmittedTimeStamp =
+                new FeedbackSubmissionSaveDraftNotificationAjaxPageData(account, timeStamp, isError);
+        
+        return createAjaxResult(lastSubmittedTimeStamp);
+    }
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/FeedbackSubmissionEditSaveDraftAction.java]: Use abstraction to write short and simple methods. Some of our `execute` methods are horribly too long. Let's try to make the new ones shorter.

**Comment by tarun14110** [comment]
@damithc @kanghj Can we do something like this ?
Info status will be displayed above the form. There will be a button at the bottom of the form 'Save Feedback'. User can anytime click that that button, the responses will be recorded. So, no loss of info if user has saved the responses. It need not be the first time (that is user hadn't submitted responses till now). User can save responses any time and continue working.
This will be like having 2 buttons at the bottom of form. One used to `submit form and stay on the page` and the other `submit form and go to homepage`. I don't think there will be any responses which user don't want to submit ? If he is writing, those are meant to submit ? And also we are giving him the functionality to edit his response anytime, so that wrong info saved shouldn't be a issue ?
In the next part, we can just remove the `save feedback` button and call ajax request periodically ?

We will save all the forms whether it be validated or not. We will just keep a flag `isFeedbackCompletelyFilled`  to track whether the form is validated or not. This will solve the problem of loosing all the responses just due to invalidated form. We will show the status to the instructor `isFeedbackCompletelyFilled`. So that he can can see who has not filled the submission completely.

As for now we are giving a Warning status if there is some non-validation in a particular question like `distribution question` and not allowing him to submit before he empty all fields or fills them correctly. What we can do is we will submit/Save all the responses except that response and displays `Please fix the error(s) for distribution question(s) 5. All other responses were successfully submitted except this response.`

**Comment by kanghj** [comment]
@tarun14110, what you are proposing sounds similar to the original draft proposal. The flag `isFeedbackCompletelyFilled` is almost identical to `isDraft` other than your proposal to show this information to instructors. But I am not sure what you are suggesting by "Save Feedback". Does that mean we submit all responses, but if validation fails for a particular response, then we save the response with `isFeedbackCompletelyFilled` as false?

**Pull request closed without merging**